File: rsome_model/Simulation.m
classdef Simulation
    properties
        t1
        t2
        net
        elec
        demands
        data_folder
        time_range
        T
        data
        vars
    end
    
    methods
        function obj = init(obj, data_folder, t1, t2)
            obj.data_folder = data_folder;

            obj.net = Network;
            obj.net = obj.net.init(data_folder);
            obj.net.data_folder;

            obj.t1 = t1;
            obj.t2 = t2;

            obj.time_range = (t1:1:t2)';
            obj.T = length(obj.time_range);
            obj.elec = obj.read_data_from_file(fullfile(data_folder, 'tariffs.csv'));
            obj.demands = obj.read_data_from_file(fullfile(data_folder, 'demands.csv'));
            obj.data = obj.build();
            obj.vars = obj.get_vars();
        end
        
        function data = read_data_from_file(obj, file_path)
            data = readtable(file_path);
        end

        function data = build(obj)
            time = obj.time_range;
            data = table(time);
            data = join(data, obj.demands, 'Keys', 'time');
            data = join(data, obj.elec, 'Keys', 'time');
            
        end

        function vars = get_vars(obj)
            facilities = unique(obj.net.fsp.facility);
            vars = table;
            for i = 1:1:size(unique(obj.net.fsp.facility), 1)
                faility_combs = obj.net.fsp(obj.net.fsp.facility == facilities(i), :);
                n_combs = height(faility_combs);

                facility_name = repmat(facilities(i),1,n_combs*obj.T)';
                time = repmat(obj.time_range, n_combs, 1);
                vars = [vars; table(time, facility_name)];
                
            end
        end

        function total_max_inflow = get_tank_max_inflow(obj, tank_idx)
                fsp_inflows = obj.net.fsp(obj.net.fsp.in == tank_idx, :);
                max_fsp_inflows = groupsummary(fsp_inflows(:, ["facility", "flow"]), "facility", "max");
                total_max_fsp_inflow = sum(table2array(max_fsp_inflows(:, 'max_flow')));

                vsp_inflows = obj.net.vsp(obj.net.vsp.in == tank_idx, :);
                max_vsp_inflows = groupsummary(vsp_inflows(:, ["name", "max_flow"]), "name", "max");
                total_max_vsp_inflow = sum(table2array(max_vsp_inflows(:, 'max_max_flow')));
                
                total_max_inflow = total_max_fsp_inflow + total_max_vsp_inflow;
        end
        
        function demand = get_tank_demand(obj, tank_idx)
                 tank_consumer = obj.net.tanks{tank_idx, "demand"};
                 demand = table2array(obj.data(:, tank_consumer));
        end
        
        function min_vol = get_min_vol_vector(obj, tank_idx, is_dynamic)
                static_min_vol = obj.net.tanks{tank_idx, "min_vol"};
                final_vol = obj.net.tanks{tank_idx, "final_vol"};

                if is_dynamic == 0
                    min_vol = static_min_vol .* ones(obj.T, 1);
                else
                    q_max = obj.get_tank_max_inflow(tank_idx);
                    demand = obj.get_tank_demand(tank_idx);
                    dynamic_min = final_vol;
                    for t=1:1:obj.T
                        v = max(static_min_vol, dynamic_min(1) + demand(t) - q_max);
                        dynamic_min = [v; dynamic_min];
                    end
                    min_vol = dynamic_min(2:end);
                end  
        end

        function vol = get_tank_vol(obj, x_fsp, x_vsp, tank_idx)
                lhs = obj.net.tanks{tank_idx, "init_vol"};
                for i = 1:1:height(obj.net.fsp.facility)
                    if obj.net.fsp{i, "in"} == tank_idx
                        mat = obj.net.get_cumulative_mat(obj.T, 1);
                        lhs = lhs + obj.net.fsp{i, "flow"} * mat * x_fsp(i, :)';
                    elseif obj.net.fsp{i, "out"} == tank_idx
                        mat = obj.net.get_cumulative_mat(obj.T, -1);
                        lhs = lhs + obj.net.fsp{i, "flow"} * mat * x_fsp(i, :)';
                    else
                        continue
                    end
                end

                for i = 1:1:height(obj.net.vsp)
                    if obj.net.vsp{i, "in"} == tank_idx
                        mat = obj.net.get_cumulative_mat(obj.T, 1);
                        lhs = lhs + mat * x_vsp(i, :)';
                    elseif obj.net.vsp{i, "out"} == tank_idx
                        mat = obj.net.get_cumulative_mat(obj.T, -1);
                        lhs = lhs + mat * x_vsp(i, :)';
                    else
                        continue
                    end
                end

                tank_consumer = obj.net.tanks{tank_idx, "demand"};
                cum_demand = table2array(cumsum(obj.data(:, tank_consumer)));
                vol = lhs - cum_demand;
        end

        function flow = get_fsp_flow(obj, x, facility)
                facility_idx = obj.net.fsp{obj.net.fsp.facility == facility, "comb"};
                faility_flows = obj.net.fsp{obj.net.fsp.facility == facility, "flow"};
                flow = faility_flows' * x(facility_idx, :);
        end

    end
end

File: simulation.py
import os
import numpy as np
import pandas as pd

from network import Network


class Simulation:
    def __init__(self, data_folder, t1, t2):
        self.data_folder = data_folder
        self.t1 = t1
        self.t2 = t2

        self.net = Network(data_folder)
        self.time_range = np.arange(t1, t2+1, 1)
        self.T = len(self.time_range)

        self.elec = pd.read_csv(os.path.join(data_folder, 'tariffs.csv'))
        self.demands = pd.read_csv(os.path.join(data_folder, 'demands.csv'))
        self.data = self.build()
        self.n_comb_facilities = len(self.net.fsp['name'].unique())

    def build(self):
        df = pd.DataFrame(index=self.time_range)
        df['hour'] = df.index.to_series() % self.T
        df = pd.merge(df, self.elec, left_on='hour', right_on='time', how='inner').drop('time', axis=1)
        df = pd.merge(df, self.demands, left_on='hour', right_on='time', how='inner').drop('time', axis=1)
        df.index = self.time_range
        return df

    def get_total_max_inflow(self, tank_idx):
        fsp_inflows = self.net.fsp.loc[self.net.fsp['in'] == tank_idx, :]
        max_fsp_inflow = fsp_inflows[['name', 'flow']].groupby('name').max()
        total_max_fsp_inflow = max_fsp_inflow['flow'].sum()

        vsp_inflows = self.net.vsp.loc[self.net.vsp['in'] == tank_idx, :]
        max_vsp_inflow = vsp_inflows[['name', 'max_flow']].groupby('name').max()
        total_max_vsp_inflow = max_vsp_inflow['max_flow'].sum()

        return total_max_fsp_inflow + total_max_vsp_inflow

    def get_tank_demand(self, tank_idx):
        tank_consumer = self.net.tanks.loc[tank_idx, "demand"]
        return self.data[tank_consumer]

    def get_min_vol_vector(self, tank_idx, is_dynamic=True):
        static_min_vol = self.net.tanks.loc[tank_idx, "min_vol"]
        final_vol = self.net.tanks.loc[tank_idx, "final_vol"]
        if not is_dynamic:
            min_vol = static_min_vol * np.ones(self.T, 1)
        else:
            q_max = self.get_total_max_inflow(tank_idx)
            demand = self.get_tank_demand(tank_idx)
            dynamic_min = [final_vol]
            for i, t in enumerate(self.time_range[::-1]):
                dynamic_min = [max(static_min_vol, dynamic_min[0] + demand.loc[t] - q_max)] + dynamic_min

            min_vol = dynamic_min[1:]
        return min_vol

    def get_tank_vol(self, tank_idx, x_fsp, x_vsp):
        mat = np.tril(np.ones((self.T, self.T)))
        tank_consumer = self.net.tanks.loc[tank_idx, 'demand']
        cum_tank_demand = self.data[tank_consumer].values.cumsum()
        lhs = self.net.tanks.loc[tank_idx, "init_vol"]

        for fsp_idx, row_fsp in self.net.fsp.iterrows():
            if row_fsp['in'] == tank_idx:
                lhs += row_fsp['flow'] * mat @ x_fsp[fsp_idx, :].T
            elif row_fsp['out'] == tank_idx:
                lhs += -1 * row_fsp['flow'] * mat @ x_fsp[fsp_idx, :].T
            else:
                continue

        for vsp_idx, row_vsp in self.net.vsp.iterrows():
            if row_vsp['in'] == tank_idx:
                lhs += mat @ x_vsp[vsp_idx, :].T
            elif row_vsp['out'] == tank_idx:
                lhs += -1 * mat @ x_vsp[vsp_idx, :].T
            else:
                continue

        vol = lhs - cum_tank_demand
        return vol

    def get_all_tanks_vol(self, x_fsp, x_vsp):
        df = pd.DataFrame()
        for tank_idx, row in self.net.tanks.iterrows():
            vol = self.get_tank_vol(tank_idx, x_fsp, x_vsp)
            df = pd.concat([df, pd.DataFrame({tank_idx: vol}, index=self.time_range)], axis=1)
        return df

    def get_facilities_flows(self, x_fsp, x_vsp):
        df = pd.DataFrame()
        for facility in self.net.fsp['name'].unique():
            facility_idx = self.net.fsp.loc[self.net.fsp['name'] == facility].index
            flows = self.net.fsp.loc[facility_idx, 'flow'].values
            facility_flow = (x_fsp[facility_idx, :] * flows[:, np.newaxis]).sum(axis=0)
            df = pd.concat([df, pd.DataFrame({facility: facility_flow}, index=self.time_range)], axis=1)

        for i, row in self.net.vsp.iterrows():
            df = pd.concat([df, pd.DataFrame({row['name']: x_vsp[i, :]}, index=self.time_range)], axis=1)

        return df

    def get_cost(self, x):
        power = self.net.fsp.loc[:, "power"].values
        power = power @ x

        cost = power * self.data.loc[:, "tariff"].values
        return cost

    def get_nominal_demands(self):
        consumers = self.net.tanks['demand']
        demands = self.data[consumers]
        demands = demands.values.flatten("F")
        return demands


